src/xpcc/driver/storage/block_device_mirror.hpp

```diff
@@ -0,0 +1,121 @@
+
+#ifndef XPCC_BLOCK_DEVICE_MIRROR_HPP
+#define XPCC_BLOCK_DEVICE_MIRROR_HPP
+
+#include <xpcc/architecture/interface/block_device.hpp>
+
+#include <xpcc/processing/resumable.hpp>
+
+namespace xpcc
+{
+
+/**
+ * \brief	Virtual block device consists of two mirrored block devices.
+ *
+ * Write operations (`erase()`, `program()` and `write()`) are forwarded
+ * to both block devices.
+ * Read operations (`read()`) are only performed on BlockDeviceA.
+ *
+ * \tparam BlockDeviceA		First block device of the mirrored block devices
+ * \tparam BlockDeviceB		Second block device
+ *
+ * \ingroup	driver_storage
+ * \author	Raphael Lehmann
+ */
+template <typename BlockDeviceA, typename BlockDeviceB>
+class BdMirror : public xpcc::BlockDevice, protected NestedResumable<3>
+{
+public:
+	/// Initializes the storage hardware
+	xpcc::ResumableResult<bool>
+	initialize();
+
+	/// Deinitializes the storage hardware
+	xpcc::ResumableResult<bool>
+	deinitialize();
+
+	/** Read data from one or more blocks
+	 *
+	 *  @param buffer	Buffer to read data into
+	 *  @param address	Address to begin reading from
+	 *  @param size		Size to read in bytes (multiple of read block size)
+	 *  @return			True on success
+	 */
+	xpcc::ResumableResult<bool>
+	read(uint8_t *buffer, uint32_t address, size_t size);
+
+	/** Program blocks with data
+	 *
+	 *  Any block has to be erased prior to being programmed
+	 *
+	 *  @param buffer	Buffer of data to write to blocks
+	 *  @param address	Address of first block to begin writing to
+	 *  @param size		Size to write in bytes (multiple of read block size)
+	 *  @return			True on success
+	 */
+	xpcc::ResumableResult<bool>
+	program(const uint8_t *buffer, uint32_t address, size_t size);
+
+	/** Erase blocks
+	 *
+	 *  The state of an erased block is undefined until it has been programmed
+	 *
+	 *  @param address	Address of block to begin erasing
+	 *  @param size		Size to erase in bytes (multiple of read block size)
+	 *  @return			True on success
+	 */
+	xpcc::ResumableResult<bool>
+	erase(uint32_t address, size_t size);
+
+	/** Writes data to one or more blocks after erasing them
+	*
+	*  The blocks are erased prior to being programmed
+	*
+	*  @param buffer	Buffer of data to write to blocks
+	*  @param address	Address of first block to begin writing to
+	*  @param size		Size to write in bytes (multiple of read block size)
+	*  @return			True on success
+	*/
+	xpcc::ResumableResult<bool>
+	write(const uint8_t *buffer, uint32_t address, size_t size);
+
+public:
+	static constexpr std::size_t BlockSizeRead = BlockDeviceA::BlockSizeRead;
+	static constexpr std::size_t BlockSizeWrite = std::max(BlockDeviceA::BlockSizeWrite, BlockDeviceB::BlockSizeWrite);
+	static constexpr std::size_t BlockSizeErase = std::max(BlockDeviceA::BlockSizeErase, BlockDeviceB::BlockSizeErase);
+
+public:
+	/** Direct access to the BlockDeviceA
+	*
+	*  @return	BlockDeviceA
+	*/
+	inline BlockDeviceA getBlockDeviceA() {return blockDeviceA;};
```

Comment:
Returning a copy is probably not what you intended. I suppose you meant BlockDeviceA&. Probably being able to copy an spi flash driver object is a bad idea. Maybe deleting the copy constructor and copy assignment operator would prevent bugs like that.